secret cache update (#471)
harness/developer-hub: docs/platform/6_Security/1-harness-secret-manager-overview.md

3. The Delegate exchanges a key pair with the Secrets Manager, over an encrypted connection.
4. The Harness Delegate uses the encrypted key and the encrypted secret and then discards them. The keys never leave the Delegate.

:::note

> Any secret manager requires a running Harness Delegate to encrypt and decrypt secrets. Any Delegate that references a secret requires direct access to the Secrets Manager.

Author: @deepakbolangady The review request was only for a specific section corresponding to secret cache. We should review the whole topic as part of a separate ticket. This topic is 2 years old and our writing style has changed since then.

In Community and Self-Managed Enterprise Edition accounts, Harness uses a random-key secrets store as the Harness Secrets Manager.

:::note
Once you have installed Self-Managed Enterprise Edition, [Add a Secrets Manager](./5-add-secrets-manager.md). By default, Self-Managed Enterprise Edition installations use the local Harness MongoDB for the default Harness Secrets Manager. This is not recommended.

:::

:::note
Harness does not currently support migrating secrets from the random-key secrets store. If you add secrets here, you will need to recreate them in any custom secrets manager you configure later.

:::

> All Harness secrets managers require a running Harness Delegate to encrypt and decrypt secrets.

Author: Removed this sentence.

If you select **Connect Through Manager**, the Harness Manager decrypts the secrets you have set up in the Harness Secrets Manager.

This is different from **Connect Through Delegate** where only the Harness Delegate, which sits in your private network, has access to your key management system.

#### Secret Cache

> In the secret manager, Harness maintains a cache of the encrypted data, except for those stored in the Hashicorp vault. The cache also contains meta-data associated with the secret like paths for the secret references, encryption key, and version. The Time To Live (TTL) for the cache is 30 minutes.

Author: Done

> Decrypting a secret sometimes requires its meta-data. When such meta-data changes on editing a secret, a cached stale value is retrieved.

Reviewer: ```suggestion
Decrypting a secret sometimes requires its metadata. When you edit a secret, its metadata changes and a stale value is retrieved from the cache.
```
Is that accurate?

> This happens in the following cases:

Reviewer: Lead-in begins with "this happens". Good to add this lead-in to the previous paragraph.

Author: Done

> - The secret manager is KMS (GCP KMS, AWS KMS).

Reviewer: ```suggestion
- The secrets manager is a KMS (GCP KMS or AWS KMS).
```
* "Secrets" for consistency. I see "secrets manager" and "secret manager". Is there a difference? (Run global check)
* Does the text in parentheses represent a set of examples? Or are they the exact set? The comma between those two names makes it unclear. If they represent an exact set, we can write "The secrets manager is Google Cloud KMS or AWS KMS." It looks like that's how those companies refer to those features, but please check. I took only a cursory glance. TBH, Google just writes "Cloud KMS", but because we aren't Google, we must prefix the company name.

Author: I checked on the usage of secrets vs secret.
AWS refers to it as AWS Secrets Manager. Google refers to it as Google Cloud Secret Manager.
I will change to "secret manager" throughout for consistency.

> - The secret manager is KMS (GCP KMS, AWS KMS): **EncryptedRecordData** contains encrypted secret values and encryption keys. When this gets cached, the cached values are always decrypted, resulting in the older value.

Reviewer 2: IMO we should not be exposing too much internals. For this point just mentioning the type of KMS should be enough.

Author: Done for all the bullets.

- The secret manager is GCP Secret Manager.

- Secret is of reference type.

> The same does not happen for secret managers who do not use KMS. The secrets for non-KMS secret managers are inline, and no meta-data is used to retrieve the value. 

Reviewer 3: "The same does not happen for secret managers who do not use KMS" I believe caching and retrieving stale value can happen for GCP Secret Manager as well when referenced secret is used and reference is changed.
@nishantsaini20 can confirm.
cc: @meenakshi-harness

Reviewer 2: @manish-manchanda  This gets covered in point **The secret manager is GCP Secret Manager** and **Secret is of reference type**

> - The secret manager is GCP Secret Manager: It contains version information in its metadata. Every time the inline secret is edited, the incremented version is updated in the database. However, the older value is retrieved because cached metadata contains information about older versions.

Reviewer: ```suggestion
- The secret manager is Google Cloud Secret Manager: It contains version information in its metadata. Every time you edit the inline secret, its version is incremented and updated in the database. However, on subsequent access, the older value is retrieved because cached metadata contains information about older versions.
```
Referring to "It contains version information in its metadata" - Please clarify what "it" is here.

> - The secret manager is GCP Secret Manager: In this case, **EncryptedRecordData** contains version information. Every time the inline secret is edited, the incremented version is updated in the database. However, because **EncryptedRecordData** is cached, and it has older version information, the older value keeps getting retrieved.

Reviewer 2: In this case, version is part of metadata. Every time the inline secret is edited, the version value also gets updated. However, because metadata is cached, and it has older version information, the older value keeps getting retrieved.

> - Secret is of reference type: If a secret's reference path is changed during editing, it will update in the database, but the cache value still holds the older path, so it will fetch the secret from the older path.

Reviewer: ```suggestion
- The secret is of the reference type: If a secret's reference path is edited, the path is updated in the database. However, the cached value continues to store the older path and fetches the stale secret.
```

> The same does not happen for secret managers who do not use KMS. The secrets for non-KMS secret managers are inline, and no meta-data is used to retrieve the value. 

Reviewer: ```suggestion
This issue does not occur with secrets managers that do not use a KMS. In such secrets managers. secrets are inline, and no metadata is used to retrieve the value. 
```
(Avoid using terms such as non-KMS. Also, I did not know what "secrets for non-KMS secret managers" meant, so I could not suggest an effective rewrite. A good way to write this is "In such secrets managers". Does that work?
